File: js/006_MaxConsecutiveOnesIII.js
/**
 * Max Consecutive Ones III
 * Given a binary array `nums` and an integer `k`, return the maximum number of 
 * consecutive 1's in the array if you can flip at most `k` 0's.
 * 
 * Link to Original Problem: https://leetcode.com/problems/max-consecutive-ones-iii/
 * 
 * @param {number[]} nums - A binary array
 * @param {number} k - The maximum number of 0's that can be flipped
 * @return {number} - The maximum number of consecutive 1's after flipping at most `k` 0's
 */
function longestOnes(nums, k)
{
    let left = 0
    let right = 0
    let zeroCount = 0
    let maxLength = 0

    while (right < nums.length)
    {
        if (nums[right] === 0)
        {
            zeroCount++
        }

        while (zeroCount > k)
        {
            if (nums[left] === 0)
            {
                zeroCount--
            }
            left++
        }

        maxLength = Math.max(maxLength, right - left + 1)
        right++
    }

    return maxLength
}

function runTests()
{
    let nums, k, expected, result

    nums = [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0]
    k = 2
    expected = 6
    result = longestOnes(nums, k)
    console.assert(result === expected, `Test 1 Failed: Expected ${expected}, got ${result}`)

    nums = [0, 0, 1, 1, 1, 0, 0]
    k = 0
    expected = 3
    result = longestOnes(nums, k)
    console.assert(result === expected, `Test 2 Failed: Expected ${expected}, got ${result}`)

    nums = [0, 0, 1, 1, 1, 0, 0, 1, 1, 1]
    k = 2
    expected = 8
    result = longestOnes(nums, k)
    console.assert(result === expected, `Test 3 Failed: Expected ${expected}, got ${result}`)

    console.log("All tests passed!")
}

runTests()

File: c#/006_MaxConsecutiveOnesIII.cs
/**
 * Max Consecutive Ones III
 * Given a binary array `nums` and an integer `k`, return the maximum number of 
 * consecutive 1's in the array if you can flip at most `k` 0's.
 * 
 * Link to Original Problem: https://leetcode.com/problems/max-consecutive-ones-iii/
 */
[TestClass]
public class MaxConsecutiveOnesIIITests
{
    public int LongestOnes(int[] nums, int k)
    {
        int left = 0, right = 0, zeroCount = 0, maxLength = 0;

        while (right < nums.Length)
        {
            if (IsZero(nums[right]))
            {
                zeroCount++;
            }

            while (zeroCount > k)
            {
                if (IsZero(nums[left]))
                {
                    zeroCount--;
                }
                left++;
            }

            maxLength = UpdateMaxLength(maxLength, right - left + 1);
            right++;
        }

        return maxLength;
    }

    private bool IsZero(int num)
    {
        return num == 0;
    }

    private int UpdateMaxLength(int currentMaxLength, int newLength)
    {
        return Math.Max(currentMaxLength, newLength);
    }

    [TestMethod]
    public void LongestOnes_ShouldReturnMaxConsecutiveOnes_WhenZerosCanBeFlipped()
    {
        var nums = new[] { 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0 };
        var k = 2;
        var expected = 6;
        var result = LongestOnes(nums, k);
        Assert.AreEqual(expected, result);
    }

    [TestMethod]
    public void LongestOnes_ShouldReturnMaxConsecutiveOnes_WhenNoZerosCanBeFlipped()
    {
        var nums = new[] { 0, 0, 1, 1, 1, 0, 0 };
        var k = 0;
        var expected = 3;
        var result = LongestOnes(nums, k);
        Assert.AreEqual(expected, result);
    }

    [TestMethod]
    public void LongestOnes_ShouldReturnMaxConsecutiveOnes_WhenSomeZerosCanBeFlipped()
    {
        var nums = new[] { 0, 0, 1, 1, 1, 0, 0, 1, 1, 1 };
        var k = 2;
        var expected = 8;
        var result = LongestOnes(nums, k);
        Assert.AreEqual(expected, result);
    }

    [TestMethod]
    public void LongestOnes_ShouldReturnArrayLength_WhenAllOnes()
    {
        var nums = new[] { 1, 1, 1, 1, 1 };
        var k = 0;
        var expected = 5;
        var result = LongestOnes(nums, k);
        Assert.AreEqual(expected, result);
    }

    [TestMethod]
    public void LongestOnes_ShouldReturnArrayLength_WhenAllZerosAndKEqualsArrayLength()
    {
        var nums = new[] { 0, 0, 0, 0, 0 };
        var k = 5;
        var expected = 5;
        var result = LongestOnes(nums, k);
        Assert.AreEqual(expected, result);
    }
}